### ISSUE TITLE
docker-machine create with amazonec2 driver doesn't properly clean up on failure

### ISSUE THREAD

**Issue by Kuzmin** [docker/machine#2968]
When trying to launch a Docker Swarm on Amazon, some of the instances will fail to come up. Docker properly detects this and gives me the full error output (usually rate limiting or issues with retrieving a gpg key, but that doesn't matter). However, it doesn't try to clean up after itself, and thus prevents me from retrying my failed command without first going onto AWS and cleaning up the keypair that `docker-machine create` left behind.

To clarify, I'm running this command

```
    docker-machine create \
        --driver amazonec2 \
        --amazonec2-access-key $AWS_ACCESS_KEY_ID \
        --amazonec2-secret-key $AWS_SECRET_ACCESS_KEY  \
        --amazonec2-vpc-id $DOCKER_SWARM_VPC_ID \
        --amazonec2-region $DOCKER_SWARM_REGION \
        --amazonec2-instance-type $DOCKER_SWARM_INSTANCE \
        --swarm \
        --swarm-discovery token://$DOCKER_SWARM_TOKEN \
        swarm-agent-01
```

This will (amongst other things, I assume) create a keypair with the name `swarm-agent-01`, and launch the instance with this keypair. If the instance comes up properly, it will be possible to remove it by running `docker-machine rm swarm-agent-01`, which will terminate the instance and remove the keypair. However, if the instance for some reason fails to deploy, it will show up in `docker-machine ls` with a status indicating failure, and terminating it with `docker-machine rm swarm-agent-01` will not remove the keypair. 

When the keypair isn't removed, trying to run the command above will fail with an error message stating `"There is already a keypair with the name swarm-agent-01.  Please either remove that keypair or use a different machine name."`. 

It would be nice if the driver would clean up after itself on failures, or at least offer to do so with a command-line argument? If the maintainers of the amazonec2 driver would accept a PR for this, I would gladly do it.

**Comment by Kuzmin** [comment]
For what it's worth, this is the GPG error I'm seeing on certain launches

```
Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.xGMTCSmiIA --no-auto-check-trustdb --trust-model always --keyring /etc/apt/trusted.gpg --primary-keyring /etc/apt/trusted.gpg --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D
+ sudo -E sh -c apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D
gpg: requesting key 2C52609D from hkp server p80.pool.sks-keyservers.net
gpg: keyserver timed out
gpg: keyserver receive failed: keyserver error
```

Maybe this is a separate bug?

**Comment by christianbundy** [comment]
@Kuzmin
It looks to me like docker/docker#13555.

**Comment by Kuzmin** [comment]
@christianbundy 
Right, yeah it looks like my comment is docker/docker#13555 

I still think the things in the opening post is a separate issue

**Comment by christianbundy** [comment]
Totally, sorry -- I should've been more clear. The original issue is absolutely a problem (that I've dealt with again and again).

**Comment by nathanleclaire** [comment]
Yeah, we should clean up after ourselves, the fundamental Machine model doesn't lend itself to it particularly well unfortunately.  I guess it's basically a huge bug if `docker-machine rm` exits 0 without having properly cleaned up or using `-f`.

**Comment by Kuzmin** [comment]
@nathanleclaire Is `docker-machine rm` really running after a failed `docker-machine create`? I only read `amazonec2.go` and couldn't find that behaviour there, but it makes sense if it's higher up. 

If so, yeah, it doesn't seem to do the right thing when there is no instance but a leftover KeyPair. It seems like it short circuits on failure to terminate instance: https://github.com/docker/machine/blob/8141874120abdc5c9056b6dbeb09bfb76a3eb290/drivers/amazonec2/amazonec2.go#L750

Maybe a more proper behavior would be to add an existence check to `Remove` and if it doesn't exist, just remove the KeyPair? Could this have some other unforeseen effects that I haven't thought about? I'm willing to submit a PR for this.

**Comment by nathanleclaire** [comment]
>  Is docker-machine rm really running after a failed docker-machine create? I only read amazonec2.go and couldn't find that behaviour there, but it makes sense if it's higher up.

If `create` fails we don't try `rm` automatically (that causes all sorts of issues).  I think this issue is related to some kind of condition where maybe `terminate` call fails, but the instance is still removed, thereby causing future `terminate` calls to also return an `error` and ensuring that the SSH key block never gets hit.  Most likely we should do both and then return a bundled error.

**Comment by nathanleclaire** [comment]
I'm working on a fix to at least try the removal of both and then bundle up the errors into one instead of bailing if we hit an error for `terminate`
